File: packages/contracts/v1/parsed-plan.ts
import type { UUID, ISO8601 } from "./common.js";
import type { BBox, Point2D, ObservationType } from "./common.js";

export interface ParsedWall {
  id: string;
  geometry: number[][];
  confidence: number;
  observation_type: ObservationType;
}

export interface ParsedOpening {
  id: string;
  geometry: number[][];
  confidence: number;
  observation_type: ObservationType;
}

export interface ParsedRoom {
  id: string;
  geometry: number[][];
  confidence: number;
  observation_type: ObservationType;
}

export interface ParsedStair {
  id: string;
  geometry: unknown;
  confidence: number;
  observation_type: ObservationType;
}

export interface ParsedColumn {
  id: string;
  geometry: unknown;
  confidence: number;
  observation_type: ObservationType;
}

export interface ParsedAnnotation {
  id: string;
  geometry: unknown;
  confidence: number;
  observation_type: ObservationType;
}

export interface AmbiguityFlag {
  id: string;
  element_type: string;
  description: string;
  confidence: number;
  element_id?: string;
}

export interface ParsedPlan {
  schema_version: "1.0";
  id: UUID;
  floorplan_asset_id: UUID;
  parse_job_id: UUID;
  status: "pending" | "partial" | "complete" | "failed";
  confidence_overall: number;
  ambiguity_flags: AmbiguityFlag[];
  walls: ParsedWall[];
  openings: ParsedOpening[];
  rooms: ParsedRoom[];
  stairs: ParsedStair[];
  columns: ParsedColumn[];
  annotations: ParsedAnnotation[];
  bounding_box_px: BBox;
  coordinate_origin: Point2D;
  created_at: ISO8601;
}

File: packages/contracts/python/noa_contracts/parsed_plan.py
from __future__ import annotations

from typing import Literal, Any
from pydantic import BaseModel

from noa_contracts.common import BBox, Point2D, ObservationType


class ParsedWall(BaseModel):
    id: str
    geometry: list[list[float]]
    confidence: float
    observation_type: ObservationType


class ParsedOpening(BaseModel):
    id: str
    geometry: list[list[float]]
    confidence: float
    observation_type: ObservationType


class ParsedRoom(BaseModel):
    id: str
    geometry: list[list[float]]
    confidence: float
    observation_type: ObservationType


class ParsedStair(BaseModel):
    id: str
    geometry: Any
    confidence: float
    observation_type: ObservationType


class ParsedColumn(BaseModel):
    id: str
    geometry: Any
    confidence: float
    observation_type: ObservationType


class ParsedAnnotation(BaseModel):
    id: str
    geometry: Any
    confidence: float
    observation_type: ObservationType


class AmbiguityFlag(BaseModel):
    id: str
    element_type: str
    description: str
    confidence: float
    element_id: str | None = None


class ParsedPlan(BaseModel):
    schema_version: Literal["1.0"] = "1.0"
    id: str
    floorplan_asset_id: str
    parse_job_id: str
    status: Literal["pending", "partial", "complete", "failed"]
    confidence_overall: float
    ambiguity_flags: list[AmbiguityFlag]
    walls: list[ParsedWall]
    openings: list[ParsedOpening]
    rooms: list[ParsedRoom]
    stairs: list[ParsedStair]
    columns: list[ParsedColumn]
    annotations: list[ParsedAnnotation]
    bounding_box_px: BBox
    coordinate_origin: Point2D
    created_at: str
